File: backend/app/api/routes/audit.py
"""Audit log routes."""

from __future__ import annotations

from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, Depends, Query

from app.api.deps import CurrentUserDep, get_audit_repository
from app.core.config import Settings, get_settings
from app.repositories.audit import AuditRepository
from app.schemas.email import AuditLogResponse, AuditSearchResponse

router = APIRouter(prefix="/audit", tags=["audit"])


@router.get(
    "",
    response_model=AuditSearchResponse,
    summary="Search audit logs by date and subject",
)
async def search_audit_logs(
    user: CurrentUserDep,
    audit_repo: Annotated[AuditRepository, Depends(get_audit_repository)],
    settings: Annotated[Settings, Depends(get_settings)],
    subject: str | None = Query(default=None),
    start_date: datetime | None = Query(default=None),
    end_date: datetime | None = Query(default=None),
    limit: int = Query(default=None, ge=1),
    cursor: str | None = Query(default=None),
) -> AuditSearchResponse:
    effective_limit = min(
        limit or settings.default_page_size,
        settings.max_page_size,
    )
    logs, next_cursor, has_more = await audit_repo.search(
        user.id,
        subject=subject,
        start_date=start_date,
        end_date=end_date,
        limit=effective_limit,
        cursor=cursor,
    )
    return AuditSearchResponse(
        count=len(logs),
        logs=[
            AuditLogResponse(
                id=log.id,
                user_id=log.user_id,
                draft_id=log.draft_id,
                email_id=log.email_id,
                action=log.action,
                subject=log.subject,
                body_snapshot=log.body_snapshot,
                created_at=log.created_at,
            )
            for log in logs
        ],
        next_cursor=next_cursor,
        has_more=has_more,
    )

File: frontend/src/lib/audit.ts
import "server-only"

import { backendFetch } from "@/src/lib/api"

export type AuditAction =
    | "approved"
    | "rejected"
    | "edited"
    | "sent"
    | "regenerated"

export type AuditLog = {
    id: string
    user_id: string
    draft_id: string | null
    email_id: string | null
    action: AuditAction
    subject: string | null
    body_snapshot: string | null
    created_at: string
}

export type AuditSearchResponse = {
    count: number
    logs: AuditLog[]
    next_cursor: string | null
    has_more: boolean
}

export async function searchAuditLogs(
    accessToken: string,
    params: {
        subject?: string
        start_date?: string
        end_date?: string
        limit?: number
        cursor?: string
    } = {},
) {
    const search = new URLSearchParams()
    if (params.subject) search.set("subject", params.subject)
    if (params.start_date) search.set("start_date", params.start_date)
    if (params.end_date) search.set("end_date", params.end_date)
    if (params.limit) search.set("limit", String(params.limit))
    if (params.cursor) search.set("cursor", params.cursor)
    const query = search.toString()
    return backendFetch<AuditSearchResponse>(
        `/api/audit${query ? `?${query}` : ""}`,
        accessToken,
    )
}
